operational/aws/shared_cost_reallocation/aws_support/shared_cost_reallocation_aws_support.pt

```diff
@@ -0,0 +1,1111 @@
+  // This is useful for debugging and monitoring the progress of the script
+  // Sampling is used to mitigate/prevent hitting the 1000 console log Policy Engine limit
+  eventLast = null // Default lastEvent to null so that the first logSample will always log
+  eventCount = 0 // Default eventCount to 0
+  function logSample(sample) {
+    delaySecondsBetweenSample = 10 // 10 seconds between log events
+    eventCount = eventCount + 1
+    var now = new Date()
+    var log_entry = "[Event "+eventCount.toString()+" @ "+now.toISOString()+"] "+JSON.stringify(sample)
+    // Check if eventLast was old than delaySecondsBetweenSample ago
+    if (eventLast == null || (now - eventLast) > delaySecondsBetweenSample*1000) {
+      eventLast = now
+    }
+  }
+
+  // Empty CSV array
+  var csv = []
+
+  var cbi_headers = _.keys(ds_cbi_optima_headers)
+  // Wrap CBI headers in quotes
+  cbi_headers_safe = _.map(cbi_headers, function(header) {
+    return '"' + header + '"'
+  })
+  csv.push(cbi_headers_safe.join(","))
+
+  var row_total = ds_cbi_csv_rows.length
+  var row_index = 0
+  var row_total_duration = 0
+  // Loop through rows
+  _.each(ds_cbi_csv_rows, function(row) {
+    // Start time for row processing
+    var startTime = new Date()
+    // Increment row index
+    row_index += 1
+    // Calculate progress percentage
+    var progress = (parseInt(row_index) / parseInt(row_total)) * 100
+    // Calculate average duration for each row
+    var avg_duration = 0
+    var avg_duration_string = ""
+    // Check if this is the first row
+    // We don't have an average duration for the first row
+    if (row_index > 1) {
+      avg_duration = row_total_duration / (row_index - 1)
+      avg_duration_string = " Average row duration: "+avg_duration.toFixed(2)+"ms"
+    }
+    logSample("Appending CSV row "+row_index.toString()+" of "+row_total.toString()+" ("+ progress.toFixed(1) +"%%)"+avg_duration_string)
+    row_values = []
+    // Loop through headers and append to row_values array
+    // This ensures that the row values are in the same order as the headers
+    _.each(cbi_headers, function(header) {
+      // Check that we have a row value for the particular CBI header
+      // Without this check the resulting value would be `"undefined"` in the CSV
+      if (typeof row[header] != "undefined") {
+        row_values.push(row[header])
+      } else {
+        // Handle the dimensions where we don't have a value from the cost data
+        row_values.push("")
+      }
+    })
+    // Wrap row values in quotes
+    row_values = _.map(row_values, function(value) {
+      return '"' + value + '"'
+    })
+    // Push row values CSV string to CSV array
+    csv.push(row_values.join(","));
+    var endTime = new Date()
+    var duration = (endTime - startTime)
+    row_total_duration += duration
+  });
+  // Join rows with newline for output CSV
+  reallocated_cost_cbi_csv = csv.join("\n")
+EOS
+end
+
+datasource "ds_cbi_create_bill_upload" do
```

Comment:
We have some CBI policies that have logic to automatically terminate hanging pending uploads before creating the new one. This solves the problem of having to use the API manually to do that if the policy ever fails mid-execution for some reason. Should that code be brought in?

Example: https://github.com/flexera-public/policy_templates/blob/7127e27352ed02f06eaf2f9e64ff98895e18764e/cost/flexera/cco/cbi_ingestion_aws_s3/cbi_ingestion_aws_s3.pt#L296

Reply:
Add in [d73b0f4](https://github.com/flexera-public/policy_templates/pull/2933/commits/d73b0f4978b3cd32bb4672cb6f61d6773299f3b3), this should be resolved